File: cpp/92_反转链表II_medium.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (head==nullptr || left >= right)
            return head;
        ListNode *dummyHead = new ListNode(0, head);
        ListNode *p = head;
        ListNode *begin = dummyHead;
        stack<ListNode*> stk;
        int pos = 1;
        while (p!=nullptr) {
            if (pos < left)
                begin = p;
            else if (pos>=left && pos<=right)
                stk.push(p);
            else 
                break;
            pos++;
            p = p->next;
        }
        // cout<<stk.top()->val;
        while (!stk.empty()) {
            begin->next = stk.top();
            stk.pop();
            begin = begin->next;
        }        
        if (begin!=nullptr)
            begin->next = p;
        return dummyHead->next;
    }
};

File: python/92_反转链表II_medium.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        dummyHead = ListNode(next = head)
        p = head
        stk = list()
        begin = dummyHead
        pos = 1
        while p:
            if pos < left:
                begin = p
            elif pos >= left and pos <= right:
                stk.append(p)
            else:
                break
            p = p.next
            pos += 1
        while stk:
            begin.next = stk.pop()
            begin = begin.next
        if begin:
            begin.next = p
        return dummyHead.next